Refactored docs (#983)
Qiskit/qiskit-experiments: docs/tutorials/custom_experiment.rst

- Implement the abstract :meth:`.BaseExperiment.circuits` method.
  This should return a list of ``QuantumCircuit`` objects defining
  the experiment payload.

- Call the :meth:`BaseExperiment.__init__` method during the subclass
  constructor with a list of physical qubits. The length of this list must
  be equal to the number of qubits in each circuit and is used to map these
  circuits to this layout during execution.
  Arguments in the constructor can be overridden so that a subclass can
  be initialized with some experiment configuration.

Optionally the following methods can also be overridden in the subclass to
allow configuring various experiment and execution options

- :meth:`.BaseExperiment._default_experiment_options`
  to set default values for configurable option parameters for the experiment.

- :meth:`.BaseExperiment._default_transpile_options`
  to set custom default values for the ``qiskit.transpile`` used to
  transpile the generated circuits before execution.

- :meth:`.BaseExperiment._default_run_options`
  to set default backend options for running the transpiled circuits on a backend.

- :meth:`.BaseExperiment._default_analysis_options`

>   to set default values for configurable options for the experiments analysis class.

Reviewer: ```suggestion
  to set default values for configurable options for the experiment's analysis class.
```

  Note that these should generally be set by overriding the :class:`.BaseAnalysis`

>   method :meth:`.BaseAnalysis._default_options` instead of this method except in the

Reviewer: ```suggestion
  method :meth:`.BaseAnalysis._default_options` instead of this method, except in the
```

>   case where the experiment requires different defaults to the used analysis class.

Reviewer: This last sentence is too advanced for a beginner. I suggest to move it to the method's doc string, and explain about it in the examples below.

Author: This was outdated anyways. I've updated the paragraph.

- :meth:`.BaseExperiment._transpiled_circuits`
  to override the default transpilation of circuits before execution.

- :meth:`.BaseExperiment._metadata`
  to add any experiment metadata to the result data.

Furthermore, some characterization and calibration experiments can be run with restless
measurements, i.e. measurements where the qubits are not reset and circuits are executed
immediately after the previous measurement. Here, the :class:`.RestlessMixin` can help
to set the appropriate run options and data processing chain.

Analysis Subclasses
~~~~~~~~~~~~~~~~~~~

To create an analysis subclass, one only needs to implement the abstract

> :meth:`.BaseAnalysis._run_analysis` method. This method takes a

Reviewer: ```suggestion
:meth:`.BaseAnalysis._run_analysis` method. This method takes an
```

:class:`.ExperimentData` container and kwarg analysis options. If any
kwargs are used the :meth:`.BaseAnalysis._default_options` method should be
overriden to define default values for these options.

The :meth:`.BaseAnalysis._run_analysis` method should return a pair

> ``(results, figures)`` where ``results`` is a list of

Reviewer: ```suggestion
``(results, figures)``, where ``results`` is a list of
```

:class:`.AnalysisResultData` and ``figures`` is a list of
:class:`matplotlib.figure.Figure`.

The :mod:`qiskit_experiments.data_processing` module contains classes for
building data processor workflows to help with advanced analysis of
experiment data.

==================================
Subclassing an Existing Experiment
==================================

This document will take you step-by-step through the process of subclassing an existing experiment in the Qiskit Experiment module.
The example in this guide focuses on adjusting the FineAmplitude experiment to calibrate on higher order transitions.
However, a similar process can be followed for other experiments.

The FineAmplitude Experiment
============================

> The ``FineAmplitude`` calibration experiment repeats N times per gate with a pulse to amplify the under-/over-rotations in the gate to determine the optimal amplitude.

Reviewer: To my understanding `FineAmplitude` is a characterization experiment, while the calibration version is `FineAmplitudeCal`.

Reviewer: @coruscating Please note this comment

Author: You're right, and I've updated the characterization experiment docstring so that it doesn't say calibration experiment anymore.

> The ``FineAmplitude`` calibration experiment repeats N times per gate with a pulse to amplify the under-/over-rotations in the gate to determine the optimal amplitude.

Reviewer: There's too many "to-verb" here. How about:
```
The ``FineAmplitude`` calibration experiment optimizes gate amplitude by repeating the gate pulse N times, hence amplifying the under-/over-rotations.
```

            return circuit

In this subclass we have overridden the ``_pre_circuit`` method in order to calibrate on higher energy transitions by using an initial X gate to populate the first excited state.

Using the Subclass
==================

Now, we can use our new subclass as we would the original parent class.
Pictured below are the results from following the Fine amplitude calibration tutorial for detecting an over-rotated pulse using our new ``HigherOrderFineXAmplitude`` class in place of the original ``FineXAmplitude`` class.
You can try this for yourself and verify that your results are similar.

.. code-block::
   
   DbAnalysisResultV1
   - name: d_theta
   - value: -0.020710672666906425 ± 0.0012903658449026907
   - χ²: 0.7819653845899581
   - quality: good
   - device_components: ['Q0']
   - verified: False

.. Writing a custom experiment
.. ===========================

.. In this tutorial, we'll use what we've learned so far to make a full experiment from
.. the :class:`.BaseExperiment` template.

.. A randomized measurement experiment
.. ===================================

.. This experiment creates a list of copies of an input circuit

> .. and randomly samples an N-qubit Paulis to apply to each one before

Reviewer: ```suggestion
.. and randomly samples an N-qubit Pauli to apply to each one before
```

> .. a final N-qubit Z-basis measurement to randomized the expected

Reviewer: ```suggestion
.. a final N-qubit Z-basis measurement to randomize the expected
```

> .. ideal output bitstring in the measured.

Reviewer: ```suggestion
.. ideal output bitstring in the measurement.
```

                circuit.add_calibration("sx", (self.physical_qubits[0],), sx_schedule, params=[])
                circuit.barrier()

            return circuit

In this subclass we have overridden the ``_pre_circuit`` method in order to calibrate on higher energy transitions by using an initial X gate to populate the first excited state.

Using the Subclass
------------------

Now, we can use our new subclass as we would the original parent class.
Shown below are results from following the :ref:`fine-amplitude-cal` tutorial
for detecting an over-rotated pulse using our new 
``HigherOrderFineXAmplitude`` class in place of the original 
:class:`.FineXAmplitude` class.
You can try this for yourself and verify that your results are similar.

.. code-block::
   
   DbAnalysisResultV1
   - name: d_theta
   - value: -0.020710672666906425 ± 0.0012903658449026907
   - χ²: 0.7819653845899581
   - quality: good
   - device_components: ['Q0']
   - verified: False

Writing a custom experiment
---------------------------

Now we'll use what we've learned so far to make a full custom experiment from
the :class:`.BaseExperiment` template.

> A randomized measurement experiment

Reviewer: Please revise the code of the experiment and analysis classes, in addition to my comments. It doesn't have the form of an example code that helps learn how to write an experiment.

Author: I've started to break up the code with explanations in-between in the newest commit. What do you mean by this doesn't have the form of an example code? Do you think a dummy template experiment with the attributes, something like

```python
class CustomExperiment(BaseExperiment):
    def __init__(self, physical_qubits=None, dummy_option=None, backend=None):
        """Initialize the experiment."""
        super().__init__(physical_qubits, analysis=CustomAnalysis(),backend=backend)
        self.set_experiment_options(
            dummy_option=dummy_option
        )
    @classmethod
    def _default_experiment_options(cls) -> Options:
        options = super()._default_experiment_options()
        options.update_options(
            dummy_option=None,
        )
        return options
    ...
```

would be helpful before introducing this experiment?

^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

This experiment creates a list of copies of an input circuit
and randomly samples an N-qubit Pauli to apply to each one before
a final N-qubit Z-basis measurement to randomize the expected

> ideal output bitstring in the measurement.

Reviewer: It's a long sentence and it has no commas. Can it be split to a sequence of shorter sentences?

The analysis uses the applied Pauli frame of a randomized
measurement experiment to de-randomize the measured counts
and combine across samples to return a single counts dictionary

> the original circuit. This has the effect of Pauli-twirling and symmetrizing the

Reviewer: same

measurement readout error. 

.. jupyter-execute::

  from numpy.random import default_rng, Generator
  from qiskit import QuantumCircuit
  from qiskit_experiments.framework import BaseExperiment

>   from qiskit.quantum_info import random_pauli_list

Reviewer: Consider arranging the imports, like pylint usually requests. Do we have pylint for the code snippets?

Author: Fixed these. I guess we can run a linter on the generated jupyter notebooks, but I don't know if we want to add that to the CI.

  class RandomizedMeasurement(BaseExperiment):
    """Randomized measurement experiment.
    """

    def __init__(
        self,
        circuit,
        measured_qubits=None,
        physical_qubits=None,
        backend=None,
        **experiment_options
    ):
        """Basic randomize Z-basis measurement via a Pauli frame transformation
        

>         Note this will just append a new set of measurment at the end of a circuit.

Reviewer: ```suggestion
        Note this will just append a new set of measurments at the end of a circuit.
```

>         A more advanced version of this experiment would be use a transpiler pass to

Reviewer: ```suggestion
        A more advanced version of this experiment would be to use a transpiler pass to
```

        replace all exisiting measurements in a circuit with randomized measurements.
        """
        if physical_qubits is None:
            physical_qubits = tuple(range(circuit.num_qubits))
        if measured_qubits is None:
            measured_qubits = tuple(range(circuit.num_qubits))
        
        # Initialize BaseExperiment
        analysis = RandomizedMeasurementAnalysis()
        super().__init__(physical_qubits, analysis=analysis, backend=backend)
        
        # Add experiment properties
        self._circuit = circuit        
        self._measured_qubits = measured_qubits
        
        # Set any init optinos
        self.set_experiment_options(**experiment_options)

    @classmethod
    def _default_experiment_options(cls):
        options = super()._default_experiment_options()
        options.num_samples = "default"
        options.seed = None
        return options
    

>     def circuits(self):

Reviewer: There's a subtle point that has to be clearly explained (perhaps it's already explained and I missed it). Naturally as a new-by user I'd expect the circuits to act on the physical qubits. Users have to understand that the circuit qubits are `0,..., len(physical_qubits)`, and a mapping is carried out later, in the transpilation.

This is also related to the explanation about parallel experiment, that says that the circuits are required to have different qubits. The accurate thing to say is that the sub-experiment must have different physical qubits.

Author: Good point, added.

        # Number of classical bits for original circuit and added measurements
        circ_nc = self._circuit.num_clbits
        meas_nc = len(self._measured_qubits)

>         circ_qubits = list(range(self.num_qubits))

Reviewer: `self.num_qubits` has not been defined here. It would be clearer to explain that it's a `BaseExperiment` property.

>         # Number of classical bits for original circuit and added measurements
>         circ_nc = self._circuit.num_clbits
>         meas_nc = len(self._measured_qubits)
>         circ_qubits = list(range(self.num_qubits))
>         circ_clbits = list(range(circ_nc))
>         meas_qubits = self._measured_qubits
>         meas_clbits = list(range(circ_nc, circ_nc + meas_nc))

Reviewer: ```suggestion
        # Number of classical bits of the original circuit
        circ_nc = self._circuit.num_clbits
        # Number of added measurements
        meas_nc = len(self._measured_qubits)
        # Circuit classical bits
        circ_clbits = list(range(circ_nc))
        # Classical bits of the added measurements
        meas_clbits = list(range(circ_nc, circ_nc + meas_nc))
        
        # Circuit qubits
        circ_qubits = list(range(self.num_qubits))
        # Qubits of the added measurements
        meas_qubits = self._measured_qubits
        
```
(rearrange)

        # Get number of samples from options
        num_samples = self.experiment_options.num_samples

>         if num_samples == "default":

Reviewer: Wouldn't it be more conventional to mark that default by setting to `None`, rather than setting to `"default"`. Also, why not just set `num_samples` to `2**self._num_qubits` in `_default_experiment_options`.

Author: I thought `_default_experiment_options` is meant for constants only? I've changed it to `None` instead of `"default"`.

            num_samples = 2 ** self.num_qubits
        
        # Get rng seed
        seed = self.experiment_options.seed
        if isinstance(seed, Generator):
            rng = seed
        else:
            rng = default_rng(seed)

        # Sample Paulis this might have duplicates, but we don't really
        # have any easy way of running different number of shots per circuit

>         # so we just run repeat circuits multiple times

Reviewer: This sentence is not clear and has grammatical errors

Author: I think this is getting too detailed anyways and isn't about how to write a custom experiment so I removed it.

        paulis = random_pauli_list(meas_nc, size=num_samples, phase=False, seed=rng)

        # Construct circuits
        circuits = []
        orig_metadata = self._circuit.metadata or {}
        for pauli in paulis:
            name = f"{self._circuit.name}_{str(pauli)}"
            circ = QuantumCircuit(
                self.num_qubits, circ_nc + meas_nc,
                name=name
            )
            # Append original circuit
            circ.compose(
                self._circuit, circ_qubits, circ_clbits, inplace=True
            )

            # Add Pauli frame
            circ.compose(pauli, meas_qubits, inplace=True)

>             # Add final Measurement

Reviewer: ```suggestion
            # Add final measurement
```

            circ.measure(meas_qubits, meas_clbits)
    

>             # Add metadata

Reviewer: Explain the role of the metadata. Explain that the analysis is going to use it, this is a way to communicate between the experiment and its analysis.

            circ.metadata = orig_metadata.copy()
            circ.metadata["rm_bits"] = meas_clbits
            circ.metadata["rm_frame"] = str(pauli)
            circ.metadata["rm_sig"] = pauli.x.astype(int).tolist()

            circuits.append(circ)

        return circuits

And the corresponding analysis class:

.. jupyter-execute::

  from qiskit_experiments.framework import BaseAnalysis, AnalysisResultData

  class RandomizedMeasurementAnalysis(BaseAnalysis):
      """Analysis for randomized measurement experiment."""

      # Helper dict to swap a clbit value
      _swap_bit = {"0": "1", "1": "0"}

>       def _run_analysis(self, experiment_data):

Author: This is explained up above in the `Analysis Subclasses` section, but I can reemphasize the point.

          
          combined_counts = {}

>           for datum in experiment_data.data():

Author: This is now explained in the getting started tutorial.

              # Get counts
              counts = datum["counts"]
              num_bits = len(next(iter(counts)))
              
              # Get metadata
              metadata = datum["metadata"]
              clbits = metadata["rm_bits"]
              sig = metadata["rm_sig"]

>               # Construct full signature

Author: Added explanation

              full_sig = num_bits * [0]
              for bit, val in zip(clbits, sig):
                  full_sig[bit] = val
              
              # Combine dicts
              for key, val in counts.items():
                  bitstring = self._swap_bitstring(key, full_sig)
                  if bitstring in combined_counts:
                      combined_counts[bitstring] += val
                  else:
                      combined_counts[bitstring] = val
                      
          
          result = AnalysisResultData("counts", combined_counts)
          return [result], []

      @classmethod

>       def _swap_bitstring(cls, bitstring, sig):

Reviewer: Not clear

.. jupyter-execute::

  from qiskit.providers.aer import AerSimulator, noise

  backend_ideal = AerSimulator()

  # Backend with asymetric readout error
  p0g1 = 0.3
  p1g0 = 0.05
  noise_model = noise.NoiseModel()
  noise_model.add_all_qubit_readout_error([[1 - p1g0, p1g0], [p0g1, 1 - p0g1]])
  noise_backend = AerSimulator(noise_model=noise_model)

We now run the experiment with a GHZ circuit on an ideal backend:

.. jupyter-execute::

  # GHZ Circuit
  nq = 4
  qc = QuantumCircuit(nq)
  qc.h(0)
  for i in range(1, nq):
      qc.cx(i-1, i)

  # Experiment parameters
  total_shots = 100000
  num_samples = 50
  shots = total_shots // num_samples

  # Run ideal randomized meas experiment
  exp = RandomizedMeasurement(qc, num_samples=num_samples)
  expdata_ideal = exp.run(AerSimulator(), shots=shots)
  counts_ideal = expdata_ideal.analysis_results("counts").value
  print(counts_ideal)

Now we repeat the experiment on the backend with readout error:

.. jupyter-execute::

  # Run noisy randomized meas experiment with readout error
  expdata_noise = exp.run(noise_backend, shots=shots)
  counts_noise = expdata_noise.analysis_results("counts").value

>   # Run noisy direct simulation of original circuit without randomization

Reviewer: ```suggestion
  # Run noisy direct simulation of the original circuit without randomization
```

>   # Run noisy direct simulation of original circuit without randomization

Reviewer: What do you mean by "direct simulation"?

Author: It means simulating the original circuit without going through the experiment. I deleted "direct" because it's confusing.

:doc:`visualization module </tutorials/visualization>` tutorial.

Custom experiment template
--------------------------

Here is a barebones template to help you get started with customization:

.. jupyter-input::

    class CustomExperiment(BaseExperiment):
        """Custom experiment class template."""

        def __init__(self, qubits=None):
            """Initialize the experiment."""
            if qubits is None:
                qubits = [0]
            super().__init__(qubits, analysis=FakeAnalysis())

        def circuits(self):
            """Generate the list of circuits to be run."""
            return []

        @classmethod
        def _default_experiment_options(cls) -> Options:
            """Update default experiment options here."""
            options = super()._default_experiment_options()
            options.update_options(
                dummy_option=None,
            )
            return options

And the corresponding analysis class template:

.. jupyter-input::

    class CustomAnalysis(BaseAnalysis):
        """Custom analysis class template."""

        def __init__(self, **kwargs):
            super().__init__()
            self._kwargs = kwargs

>         def _run_analysis(self, experiment_data, **options):

Author: My bad, I copied this from `FakeAnalysis` without checking carefully (that should probably be updated too). I've updated both templates.

a randomized measurement experiment to de-randomize the measured counts. The results
are then combined across samples to return a single counts dictionary for
the original circuit. This has the effect of Pauli twirling and symmetrizing the
measurement readout error.

To start, we write our own ``__init__()`` method to take as input the circuit that
we want to twirl on. We also want to give the user the option to specify which
physical qubits to run the circuit over, and which qubits to measure over. If the user
doesn't specify these options, we default both to the list of qubits starting with 0 with
the length of the number of qubits in the circuit for both.

.. jupyter-input::

    from numpy.random import default_rng, Generator
    from qiskit import QuantumCircuit
    from qiskit.quantum_info import random_pauli_list
    from qiskit_experiments.framework import BaseExperiment

    class RandomizedMeasurement(BaseExperiment):
    """Randomized measurement experiment."""
        def __init__(
            self,
            circuit,
            measured_qubits=None,
            physical_qubits=None,
            backend=None,

>             **experiment_options

Author: Done.

be able to set how many repetitions of the circuit to run:

.. jupyter-input::

    ...

        @classmethod
        def _default_experiment_options(cls):
            options = super()._default_experiment_options()
            options.num_samples = None
            options.seed = None
            return options

Now we write the ``circuits()`` method. We need to take the input circuit in
``self._circuit`` and add our random Paulis as well as measurement at the end. We use
the built-in property :attr:`~.BaseExperiment.num_qubits` of :class:`~.BaseExperiment`
to get the number of qubits in the experiment. We keep track of the list of qubits and
classical Note that the circuits themselves are always built on qubits `0` to `length of
the circuit - 1`, and not the actual physical qubit indices given in
``physical_qubits``, as discussed in :doc:`getting_started`.

.. jupyter-input::

    ...

        def circuits(self):
            # Number of classical bits of the original circuit
            circ_nc = self._circuit.num_clbits

            # Number of added measurements
            meas_nc = len(self._measured_qubits)

            # Classical bits of the circuit

>             circ_clbits = list(range(circ_nc))

Reviewer 2: Maybe somewhere you want to mention that the input circuit could mid-circuit measurements? Or maybe for the sake of the example it is cleaner to assume no extra measurements and the measurements added by the experiment are the only measurements?

Author: I added a simple mid-circuit measurement circuit at the end to demonstrate this, though I don't run it. Maybe there's a more interesting example to actually run?

            circ.compose(pauli, meas_qubits, inplace=True)

            # Add final measurement
            circ.measure(meas_qubits, meas_clbits)

            circ.metadata = orig_metadata.copy()
            circ.metadata["rm_bits"] = meas_clbits
            circ.metadata["rm_frame"] = str(pauli)
            circ.metadata["rm_sig"] = pauli.x.astype(int).tolist()
            circuits.append(circ)
        return circuits

Now we write the analysis class, overriding ``_run_analysis`` as described above. We
loop over each circuit to process the output bitstring. Since we're using default level 
2 data, we access it with the ``counts`` key. We use the circuit metadata to calculate the bitwise XOR mask from the Pauli
signature to restore the output to what it should be without the random Pauli frame
at the end. We make a new :class:`.AnalysisResultData` object since we're rewriting the 
counts from the original experiment.

.. jupyter-input::

    from qiskit_experiments.framework import BaseAnalysis, AnalysisResultData

    class RandomizedMeasurementAnalysis(BaseAnalysis):
        """Analysis for randomized measurement experiment."""

        # Helper dict to swap a clbit value

>         _swap_bit = {"0": "1", "1": "0"}

Author: I see your point but I also think it's fine to not redefine it every time `_swap_bitstring` is called. I did move it to the  `_swap_bitstring` section so it's easier to follow.

    counts_ideal = expdata_ideal.analysis_results("counts").value
    print(counts_ideal)

Repeat the experiment on the backend with readout error and compare with results
from running GHZ circuit itself:

.. jupyter-execute::

    # Run noisy randomized meas experiment with readout error
    expdata_noise = exp.run(noise_backend, shots=shots)
    counts_noise = expdata_noise.analysis_results("counts").value

    # Run noisy simulation of the original circuit without randomization
    meas_circ = qc.copy()
    meas_circ.measure_all()
    result = noise_backend.run(meas_circ, shots=total_shots).result()
    counts_direct = result.get_counts(0)

    from qiskit.visualization import plot_histogram

    # Plot counts, ideally randomized one should be more symmetric in noise
    # than direct one with asymmetric readout error
    plot_histogram([counts_ideal, counts_direct, counts_noise],
                legend=["Ideal",
                        "Asymmetric meas error (Direct)",
                        "Asymmetric meas error (Randomized)"])

The asymmetric measurement of the original circuit on this backend (Direct on the plot legend)

> has been successfully symmetrized by the application of randomized measurement 

Reviewer 2: Maybe you want a comment saying that for the GHZ state symmetrizing the error should produce a symmetric graph? I don't think that is true for an arbitrary input circuit.

Author: Added a sentence.
